warehouse/slave.go

```diff
@@ -189,126 +207,170 @@ func (job *Payload) getDiscardsTable() string {
+				jr.job.LoadFilePrefix,
+			)
+		} else {
+			uploadLocation, err = uploader.Upload(
+				ctx,
+				file,
+				jr.loadObjectFolder,
+				tableName,
+				jr.job.SourceID,
+				getBucketFolder(jr.job.UniqueLoadGenID, tableName),
+			)
+		}
+		return uploadLocation, err
+	}
+
+	process := func(uploadJobChan <-chan *uploadJob) <-chan *result {
+		processStream := make(chan *result, len(jr.outputFileWritersMap))
+
+		g, _ := errgroup.WithContext(ctx)
+		g.SetLimit(jr.numLoadFileUploadWorkers)
+
+		for i := 0; i < jr.numLoadFileUploadWorkers; i++ {
+			g.Go(func() error {
+				for uploadJob := range uploadJobChan {
+					select {
+					case <-ctx.Done():
+						return ctx.Err()
+					default:
+						loadFileUploadStart := time.Now()
+
+						uploadOutput, err := uploadLoadFile(
+							uploader,
+							uploadJob.outputFile,
+							uploadJob.tableName,
+						)
+						if err != nil {
+							return fmt.Errorf("uploading load file: %w", err)
+						}
 
-	// Wait for response
-	for {
-		select {
-		case loadFileOutput := <-loadFileOutputChan:
-			loadFileUploadOutputs = append(loadFileUploadOutputs, loadFileOutput)
-			if len(loadFileUploadOutputs) == len(jobRun.outputFileWritersMap) {
-				return loadFileUploadOutputs, nil
+						mu.Lock()
+						totalUploadTime += jr.since(loadFileUploadStart)
+						mu.Unlock()
+
+						loadFileStats, err := os.Stat(uploadJob.outputFile.GetLoadFile().Name())
+						if err != nil {
+							return fmt.Errorf("getting load file stats: %w", err)
+						}
+
+						processStream <- &result{
+							uploadOutput: loadFileUploadOutput{
+								TableName:             uploadJob.tableName,
+								Location:              uploadOutput.Location,
+								ContentLength:         loadFileStats.Size(),
+								TotalRows:             jr.tableEventCountMap[uploadJob.tableName],
+								StagingFileID:         jr.job.StagingFileID,
+								DestinationRevisionID: jr.job.DestinationRevisionID,
+								UseRudderStorage:      jr.job.UseRudderStorage,
+							},
+						}
+					}
+				}
+				return nil
+			})
+		}
+
+		go func() {
+			if err := g.Wait(); err != nil {
+				processStream <- &result{err: err}
 			}
-		case err := <-uploadErrorChan:
-			pkgLogger.Errorf("received error while uploading load file to bucket for staging file ids %s, cancelling the context: err %v", stagingFileId, err)
-			return []loadFileUploadOutput{}, err
-		case <-time.After(slaveUploadTimeout):
-			return []loadFileUploadOutput{}, fmt.Errorf("load files upload timed out for staging file idsx: %v", stagingFileId)
+			close(processStream)
+		}()
+
+		return processStream
+	}
+
+	for processedJob := range process(generate()) {
+		if err := processedJob.err; err != nil {
+			return nil, fmt.Errorf("uploading load file to object storage: %w", err)
 		}
+
+		output = append(output, processedJob.uploadOutput)
+	}
+	if ctx.Err() != nil {
+		return nil, fmt.Errorf("uploading load file to object storage: %w", ctx.Err())
 	}
```

Comment:
The upload has already been done though, there is no computation left on our side here. I think we can remove this check and just return. The context can be checked upstream as needed, right?

Reply:
Thanks for pointing out. It's not required.